### PR TITLE
feat: add copy_to_directory_bin_action with improved performance by moving glob match & copying to a golang binary

### PR DESCRIPTION
**Pull request by gregmagolan**
This is a pre-factor for improved performance on `copy_to_directory` and `npm_package` (in rules_js) which will address https://github.com/aspect-build/bazel-lib/issues/301.

Not yet load bearing on `copy_to_directory` but ready for review.

In a follow-up I will add release artifacts for the copy_to_directory golang binary in https://github.com/aspect-build/bazel-lib/pull/309 & then another with one with a toolchain definition. Those should land first to get a release of the binaries out before the `copy_to_directory` flip can happen.

### CODE REVIEW

**Comment by gregmagolan** [comment]
BREAKING CHANGES that will come:
```
$ gd lib/tests/copy_to_directory/BUILD.bazel
diff --git a/lib/tests/copy_to_directory/BUILD.bazel b/lib/tests/copy_to_directory/BUILD.bazel
index 66c289d..21b4577 100644
--- a/lib/tests/copy_to_directory/BUILD.bazel
+++ b/lib/tests/copy_to_directory/BUILD.bazel
@@ -82,10 +82,8 @@ copy_to_directory(
         # flatten test_a & test_b to the root
         "test_a": "",
         "test_b": "",
-        # some paths that won't match
-        "a/": "wont_match_a_is_terminal_path",
-        "a/a2": "wont_match_since_a2_is_in_a_tree_artifact",
     },
+    verbose = True,
 )
 
 diff_test(
@@ -178,8 +176,8 @@ copy_to_directory(
     name = "case_8",
     srcs = case_srcs,
     exclude_srcs_patterns = [
-        "a",
-        "b",
+        "a/*",
+        "b/*",
         "e/e*",
     ],
 )
@@ -224,8 +222,6 @@ copy_to_directory(
         # flatten test_a & test_b to the root
         "test_a": "",
         "test_b": "",
-        # some paths that won't match
-        "a/a2": "wont_match_since_a2_is_not_included_in_the_directory_path_in_srcs",
     },
 )
 
@@ -313,8 +309,8 @@ copy_to_directory(
     name = "case_13",
     srcs = case_srcs,
     include_srcs_patterns = [
-        "a",
-        "b",
+        "a/*",
+        "b/*",
         "?/e*",
         "f/f2/f2",
     ],
@@ -334,8 +330,8 @@ copy_to_directory(
         "f/f2/f1",
     ],
     include_srcs_patterns = [
-        "a",
-        "b",
+        "a/*",
+        "b/*",
         "?/e*",
         "f/f?/**",
     ],
```

**Comment by gregmagolan** [comment]
Note to self for the flip:
```
    """Legacy helper function to copy files to a directory.

    This helper calculates copy paths in Starlark during analysis and performs the copies in a
    bash/bat script. For improved analysis and runtime performance, it is recommended the switch
    to `copy_to_directory_bin_action` which calculates copy paths and performs copies with a tool
    binary, typically the `@aspect_bazel_lib//tools/copy_to_directory` `go_binary` either built
    from source or provided by a toolchain.

    This helper is used by copy_to_directory. It is exposed as a public API so it can be used within
    other rule implementations where additional_files can also be passed in.
```

File: WORKSPACE [hunk 1]
```diff
@@ -29,11 +29,17 @@ load("//lib:host_repo.bzl", "host_repo")
 
 host_repo(name = "aspect_bazel_lib_host")
 
-load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies")
+load("//:deps.bzl", "go_dependencies")
+
+# gazelle:repository_macro deps.bzl%go_dependencies
+# gazelle:repository go_repository name=org_golang_x_tools importpath=golang.org/x/tools
+# https://github.com/bazelbuild/bazel-gazelle/issues/1217#issuecomment-1152236735
+go_dependencies()
```
Comment (kormide) [inline WORKSPACE:37]: We'll also need to call this in the bzlmod module extension.

Comment (gregmagolan) [reply WORKSPACE:37]: This dep won't be passed on to users. They'll be a toolchain to setup instead similar to the yq & jq toolchains.

**Review by kormide — APPROVED** [review]

File: tools/copy_to_directory/main.go [hunk 2]
```diff
@@ -1,7 +1,347 @@
...
+	}
+
+	// Calculate copy paths
+	copyPaths, err := calcCopyPaths(cfg)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Perform copies
+	// TODO: split out into parallel go routines?
```
Comment (kzadorozhny) [inline tools/copy_to_directory/main.go:333]: Yes. The `calcCopyPaths` could also benefit from parallelization. `copy` operation can be scheduled as soon as a file is found.

Comment (gregmagolan) [reply tools/copy_to_directory/main.go:333]: Good idea. We can do some fancy stuff with it in golang. Tho I may wait until @f0rmiga is back from vacation in a week before tackling it as he's much better at go routines than I am.

**Pull request merged**